# Session 4

In this session we will run through the basic steps for analysing a simply RNA-seq experiment. This includes:

* filtering out lowly expressed genes
* normalisation
* creating a multidimensional scaling plot (MDS)
* creating a design matrix
* fitting gene-wise linear models (with empirical Bayes moderation to more accurately estimate gene-wise variability) 
* performing statistical testing for differential expression

The aim of this session is to give you an basic understanding and intuition about each of the basic RNA-seq analysis steps. It does NOT aim to teach you statistics. 


## Filtering

Let's start by re-creating our `DGEList` object:

```{r, warning=FALSE}
# load required packages
library(edgeR)
library(limma)

# vector of file names
files <- c("GSM1545535_10_6_5_11.txt", "GSM1545536_9_6_5_11.txt", 
   "GSM1545538_purep53.txt", "GSM1545539_JMS8-2.txt", 
   "GSM1545540_JMS8-3.txt", "GSM1545541_JMS8-4.txt", 
   "GSM1545542_JMS8-5.txt", "GSM1545544_JMS9-P7c.txt", 
   "GSM1545545_JMS9-P8c.txt")

# creage DGEList object
x <- readDGE(files, path = "data", columns = c(1,3))

# add sample names
samplenames <- c("10_6_5_11", "9_6_5_11", "purep53", "JMS8-2", "JMS8-3",
                 "JMS8-4", "JMS8-5", "JMS9-P7c", "JMS9-P8c")

rownames(x$samples) <- samplenames

colnames(x$counts) <- samplenames

# add phenotype group information
group <- as.factor(c("LP", "ML", "Basal", "Basal", "ML", "LP", "Basal", "ML", "LP")) 
x$samples$group <- group

# add gene annotation information
geneAnnot <- read.delim("data/Ses3_geneAnnot.tsv", stringsAsFactors = FALSE)
x$genes <- geneAnnot
```

The first step is to filter out lowly expressed genes. There are two main problems with low abundant genes:

* Technical variation is more problematic for low abundance genes. This variation is thought to be due to two factors; insufficient mixing and low sampling fraction [@mcintyre2011rna]. 
    * Insufficient mixing of solutions during library preparation can result in uneven distribution of reads. 
    * RNA sequencing can be thought of as sampling. You start with many, many RNA molecules. You sample from this 'population' of RNAs by sequencing a small proportion of the RNA molecules you started out with. Your sample consists of the RNA molecules that were sequenced. This is generally a very small proportion of the total 'population' of RNAs you started with. For example, if you start with 100ng of mRNA and sequence ~300 million reads, you only sequence 0.0013% of the mRNA that you started with [@mcintyre2011rna]. It is difficult to obtain accurate measures of the true abundance of each RNA species when you only sample such a small proportion of the total 'population' of mRNAs. Measurement errors will occur simply due to the random nature of the sampling process. This problem affects lowly abundant RNA species more because the proportion of the error relative to the (small) count value for that RNA species, is larger (than it would be for more highly abundant RNA species).
* Biologically, genes that are expressed at low, biologicaly not meaningful, levels are not of interest.

Additionally, genes that have a count of 0 across ALL samples are also removed, simply because they provide no useful information.

Removing these highly variable, lowly expressed genes increases your 'power' to detect differentially expressed genes [@bourgon2010independent]. To test for differential expression, a statistical test must be conducted for every gene. There are a large number of genes - meaning that a large number of statistical tests must be performed. When you perform many statistical tests, many will produce small p-values by chance alone. We must account for these false positives. There are a number of methods to do this and these are broadly called 'adjusting for multiple testing'. This adjustment reduces the about of false positives but comes at the cost of reduced 'power' to detect true positives. If we filter out uninteresting, lowly expressed genes, we need to perform fewer statistical tests and reduce the impact that multiple testing adjustment has on detection power. 

The `filterByExpr()` function provides an automatic way to filter genes.

By default, it keeps genes with a count of 10 or more in a minimum number of samples. This minimum number of samples is the smallest phenotype group. In our experiment there are 3 phenotype groups each with 3 samples. Filtering is thus performed such that genes that remain are genes with a count of 10 in at least 3 samples.

Note that the actual filtering is done using counts per million, so there is no bias against samples with small library sizes.

The output of this function is a vector of logicals, indicating which rows (genes) should be kept and which filtered.

```{r}
keep <- filterByExpr(x, group = group)
head(keep)
```


```{r}
x <- x[keep,, keep.lib.sizes=FALSE]
str(x$counts)
```

We can see that we now have 16624 genes. We started with 27179 genes - meaning that ~40% of genes have been filtered out.

## Normalisation

The aim of normalisation is to remove systematic technical effects. There are two main factors that need to be normalised for in RNA-seq:

* Sequencing depth/library size - technically sequencing a sample to half the depth will give, on average, half the number of reads mapping to each gene [@robinson2010scaling].
* RNA composition - if a large number of genes are unique to, or highly expressed, in one experimental conditon, the sequencing 'read estate' available for the remaining genes in that sample is decreased. For example, let's say we have 2 samples, A and B. Both contain the same set of genes which are also expressed at the same level. In sample A however, there are also many genes that are not expressed in sample B. Each sample is sequenced to the same depth. The genes that are present and actually expressed equally in the two samples will have fewer sequencing reads in sample A. This is because the genes present in sample A but NOT present in sample B are taking up a portion of the sequencing 'real estate'. As a result there is less reads 'available' to the other genes (the ones present in both samples). This means that the 'proportion o reads attributed to a given gene in a library depends on the expression properties of the whole sample rather than just the expression of that gene [@robinson2010scaling]. 

Sequencing depth is accounted for by calculating the counts per million (cpm). This metric is calculated by:

1. taking the library size (sum of all counts for a sample), 
2. dividing this by 1,000,000 to get the 'per million' scaling factor,
3. then dividing all read counts for each gene in that sample by the 'per million' scaling factor

RNA composition can be accounted for by using more sophisticated normalisation methodologies. We will use 'trimmed mean of M-values' (TMM), whcih estilates relative RNA levels from RNA-seq data [@robinson2010scaling]. TMM calculates a library size scaling factor for each library (sample). This is done using the following steps:

1. calculate the gene expression log fold changes and absolute expression values for pair-wise samples (selecting one sample from the experiment as a reference)
2. remove the genes with the highest and lowest fold changes and absolute expression values
3. take a weighted mean of the remaining genes (where the weight is the inverse of the approximate asymptotic variances). This gives the normalisation factor for each library (sample)

TMM, like most normalisation methods, assumes that the majority of genes are NOT differentially expressed.

Subsequent steps in this analysis will use log-cpm values, calculated using the normalisation factors, which 'scale' each library size.

We can calculate the normalisation factors, specifying that we want to use the `"TMM"` method:

```{r}
x <- calcNormFactors(x, method = "TMM")
```

This function calculates the normalisation factors for each library (sample) and puts this information in the `samples` dataframe. Note that it takes x (our `DGEList` object as input) and returns a `DGEList` object as well.

Let's take a look at our normalisation factors:

```{r}
x$samples$norm.factors
```

These normalisation factors are all close to 1 for all samples, suggesting minimal difference in 'relative RNA levels' between samples.

## MDS plot

MDS is a dimension reduction technique. Given a matrix of distances between each pair of objects (each pair of samples in our case), it tries to place each object in a lower dimensional space (usually 2 dimensions, such that a scatter plot is created) such that the distances between each object (sample) is preserved as well as possible.

>Note on dimension reduction:
>
>Dimension reduction is a process of reducing the number of 'variables' under consideration. There are a number of reasons why we may want to do this. In RNA-seq experiments specifically, the main advantage of dimension reduction is to better visualise and thus explore your data. Take for example, a gene expression experiment where the expression level of only 2 genes are measured in a number of samples. We can easily visualise this by plotting a scatter graph where gene 1 expression is on the y axis and gene 2 expression is on the x axis (or vice versa) and each point is a sample. In RNA-seq experiments however, we generally measure the expression of over 10,000 genes. It would be very difficult to visualise such data as we would need to have over 10,000 axis (as we have over 10,000 genes). What we want is to reduce the number of dimensions (genes, in this case) such that we can visualise the data on a plot in 2 dimensions. This is the idea behind dimension reduction processes like principal components analysis (PCA) and MDS.



**Challenge 4.1**




## Linear modelling





### Design matrix




## Statistical testing




## Answers

**Challenge 4.1**









