# Session 4
```{r}
library(knitr)
library(kableExtra)
```

In this session we will run through the basic steps for analysing a simply RNA-seq experiment. This includes:

* filtering out lowly expressed genes
* normalisation
* creating a multidimensional scaling (MDS) plot 
* creating a design matrix
* fitting gene-wise linear models (with empirical Bayes moderation to more accurately estimate gene-wise variability) 
* performing statistical testing for differential expression

The aim of this session is to give you an basic understanding and intuition about each of the basic RNA-seq analysis steps. It does NOT aim to teach you statistics. 


## Filtering

Let's start by re-creating our `DGEList` object:

```{r, warning=FALSE}
# load required packages
library(edgeR)
library(limma)

# vector of file names
files <- c("GSM1545535_10_6_5_11.txt", "GSM1545536_9_6_5_11.txt", 
   "GSM1545538_purep53.txt", "GSM1545539_JMS8-2.txt", 
   "GSM1545540_JMS8-3.txt", "GSM1545541_JMS8-4.txt", 
   "GSM1545542_JMS8-5.txt", "GSM1545544_JMS9-P7c.txt", 
   "GSM1545545_JMS9-P8c.txt")

# creage DGEList object
x <- readDGE(files, path = "data", columns = c(1,3))

# add sample names
samplenames <- c("10_6_5_11", "9_6_5_11", "purep53", "JMS8-2", "JMS8-3",
                 "JMS8-4", "JMS8-5", "JMS9-P7c", "JMS9-P8c")

rownames(x$samples) <- samplenames

colnames(x$counts) <- samplenames

# add phenotype group information
group <- as.factor(c("LP", "ML", "Basal", "Basal", "ML", "LP", "Basal", "ML", "LP")) 
x$samples$group <- group

# add gene annotation information
geneAnnot <- read.delim("data/Ses3_geneAnnot.tsv", stringsAsFactors = FALSE)
x$genes <- geneAnnot
```

The first step is to filter out lowly expressed genes. There are two main problems with low abundant genes:

* Technical variation is more problematic for low abundance genes. This variation is thought to be due to two factors; insufficient mixing and low sampling fraction [@mcintyre2011rna]. 
    * Insufficient mixing of solutions during library preparation can result in uneven distribution of reads. 
    * RNA sequencing can be thought of as sampling. You start with many, many RNA molecules. You sample from this 'population' of RNAs by sequencing a small proportion of the RNA molecules you started out with. Your sample consists of the RNA molecules that were sequenced. This is generally a very small proportion of the total 'population' of RNAs you started with. For example, if you start with 100ng of mRNA and sequence ~300 million reads, you only sequence 0.0013% of the mRNA that you started with [@mcintyre2011rna]. It is difficult to obtain accurate measures of the true abundance of each RNA species when you only sample such a small proportion of the total 'population' of mRNAs. Measurement errors will occur simply due to the random nature of the sampling process. This problem affects lowly abundant RNA species more because the proportion of the error relative to the (small) count value for that RNA species, is larger (than it would be for more highly abundant RNA species).
* Biologically, genes that are expressed at low, biologicaly not meaningful, levels are not of interest.

Additionally, genes that have a count of 0 across ALL samples are also removed, simply because they provide no useful information.

Removing these highly variable, lowly expressed genes increases your 'power' to detect differentially expressed genes [@bourgon2010independent]. To test for differential expression, a statistical test must be conducted for every gene. There are a large number of genes - meaning that a large number of statistical tests must be performed. When you perform many statistical tests, many will produce small p-values by chance alone. We must account for these false positives. There are a number of methods to do this and these are broadly called 'adjusting for multiple testing'. This adjustment reduces the about of false positives but comes at the cost of reduced 'power' to detect true positives. If we filter out uninteresting, lowly expressed genes, we need to perform fewer statistical tests and reduce the impact that multiple testing adjustment has on detection power. 

The `filterByExpr()` function provides an automatic way to filter genes.

By default, it keeps genes with a count of 10 or more in a minimum number of samples. This minimum number of samples is the smallest phenotype group. In our experiment there are 3 phenotype groups each with 3 samples. Filtering is thus performed such that genes that remain are genes with a count of 10 in at least 3 samples.

Note that the actual filtering is done using counts per million, so there is no bias against samples with small library sizes.

The output of this function is a vector of logicals, indicating which rows (genes) should be kept and which filtered.

```{r}
keep <- filterByExpr(x, group = group)
head(keep)
```


```{r}
x <- x[keep,, keep.lib.sizes=FALSE]
str(x$counts)
```

We can see that we now have 16624 genes. We started with 27179 genes - meaning that ~40% of genes have been filtered out.

## Normalisation

The aim of normalisation is to remove systematic technical effects. There are two main factors that need to be normalised for in RNA-seq:

* Sequencing depth/library size - technically sequencing a sample to half the depth will give, on average, half the number of reads mapping to each gene [@robinson2010scaling].
* RNA composition - if a large number of genes are unique to, or highly expressed, in one experimental conditon, the sequencing 'read estate' available for the remaining genes in that sample is decreased. For example, let's say we have 2 samples, A and B. Both contain the same set of genes which are also expressed at the same level. In sample A however, there are also many genes that are not expressed in sample B. Each sample is sequenced to the same depth. The genes that are present and actually expressed equally in the two samples will have fewer sequencing reads in sample A. This is because the genes present in sample A but NOT present in sample B are taking up a portion of the sequencing 'real estate'. As a result there is less reads 'available' to the other genes (the ones present in both samples). This means that the 'proportion o reads attributed to a given gene in a library depends on the expression properties of the whole sample rather than just the expression of that gene [@robinson2010scaling]. 

Sequencing depth is accounted for by calculating the counts per million (cpm). This metric is calculated by:

1. taking the library size (sum of all counts for a sample), 
2. dividing this by 1,000,000 to get the 'per million' scaling factor,
3. then dividing all read counts for each gene in that sample by the 'per million' scaling factor

RNA composition can be accounted for by using more sophisticated normalisation methodologies. We will use 'trimmed mean of M-values' (TMM), whcih estilates relative RNA levels from RNA-seq data [@robinson2010scaling]. TMM calculates a library size scaling factor for each library (sample). This is done using the following steps:

1. calculate the gene expression log fold changes and absolute expression values for pair-wise samples (selecting one sample from the experiment as a reference)
2. remove the genes with the highest and lowest fold changes and absolute expression values
3. take a weighted mean of the remaining genes (where the weight is the inverse of the approximate asymptotic variances). This gives the normalisation factor for each library (sample)

TMM, like most normalisation methods, assumes that the majority of genes are NOT differentially expressed.

Subsequent steps in this analysis will use log-cpm values, calculated using the normalisation factors, which 'scale' each library size.

We can calculate the normalisation factors, specifying that we want to use the `"TMM"` method:

```{r}
x <- calcNormFactors(x, method = "TMM")
```

This function calculates the normalisation factors for each library (sample) and puts this information in the `samples` dataframe. Note that it takes x (our `DGEList` object as input) and returns a `DGEList` object as well.

Let's take a look at our normalisation factors:

```{r}
x$samples$norm.factors
```

These normalisation factors are all close to 1 for all samples, suggesting minimal difference in 'relative RNA levels' between samples.

## MDS

MDS is a way to visualise distances between sets of objects (samples in our case). It is also a dimension reduction technique, like principal components analysis (PCA). Given a matrix of distances between each pair of objects (samples in our case), it tries to place each object in a lower dimensional space (usually 2 dimensions, so a scatter plot can be created) such that the distances between each object (sample) is preserved as well as possible.

>Note on dimension reduction:
>
>Dimension reduction is a process of reducing the number of 'variables' under consideration. There are a number of reasons why we may want to do this. In RNA-seq experiments specifically, the main advantage of dimension reduction is to allow visualisation and exploration your data. Take for example, a gene expression experiment where the expression level of only 2 genes are measured in a number of samples. We can easily visualise this by plotting a scatter graph where gene 1 expression is on the y axis and gene 2 expression is on the x axis (or vice versa) and each point is a sample. In RNA-seq experiments however, we generally measure the expression of over 10,000 genes. It would be impossible to visualise such data as we would need to have over 10,000 axes (as we have over 10,000 genes) in our graph. What we want, is to reduce the number of dimensions such that we can visualise the data on a plot in 2 or 3 dimensions. Information is inevitably lost during the dimension reduction process, though all methodologies aim to preserve as much information as possible. 

The `plotMDS()` function performs MDS on a `DGEList` object using metric MDS. The distances between each point (sample) in the resulting plot can be interpreted as the typical log2-fold-change between the samples, for the most different genes (genes that have the most different expression values between that pair of samples).

Let's take a look at the plot created from `x`, our `DGEList` object:

```{r}
plotMDS(x)
```

Each point on the plot represents one sample and is 'labelled' using the sample name.

We can change the labelling to use the name of the group the sample belongs to instead. Recall `group` is a vector that, specifies the phenotype group of each sample (in the order the samples are in within `x`):

```{r}
group
```

To change the labels we can simply use:

```{r}
plotMDS(x, labels = group)
```

This shows us that the phenotype groups tend to 'cluster' together, meaning that the gene expression profiles are similar for samples within a phenotype group. The 'Basal' type samples quite close together whiel the 'LP' (luminal progenitor) and 'ML' (mature luminal) type samples are further apart, signifying that their expression profiles are more variable.

To make the three phenotype groups more distinct in our graph, we are going to colour samples from each group differently. To do this, we will use the `col` argument in `plotMDS()`. `col` takes in a vector the same length as the number of points in the plot (9 in our case, as there are 9 samples). Each element of the vector should be a colour name (R understands [over 600 colour names](http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf)), indicating what colour that sample should be.

To make this more clear, take a look at the table below, which lists all the samples and the phenotype group they belong to:

```{r, echo=FALSE}
kable(
  data.frame(Samples = rownames(x$samples), Group = x$samples$group)
) %>%
  kable_styling(full_width = F)
```

For example, let's say we wanted LP samples to be coloured green, ML samples to be coloured red and Basal samples to be coloured blue. The `col` argument would then require a vector that looks like this:

```{r, echo=FALSE, eval=TRUE}
groupCol <- group
levels(groupCol) <- c("blue", "green", "red")
groupCol <- as.character(groupCol)
groupCol
```

Of course, we dont want to type out this vector ourselves. We want to make R do this. This is less tedious and error prone.

We can do this using the `group` vector. Recall this factor vector gives the phenotype groups of each sample (in the same order as the samples are in within `x`, our `DGEList` object).

```{r}
group
```

We don't want to change this vector, so we will create a new vector called `groupCol`, that is the same as `group`:

```{r}
groupCol <- group
```

We can now create our character vector. I will guide you through this in the next challenge.

**Challenge 4.1**

1. The `levels` function gives all the 'categories' of a factor vector. If we run it on `groupCol`, it will give our 3 phenotypes:

```{r}
levels(groupCol)
```

Replace the current levels (categories) of `groupCol` with a character vector containing 3 colour names of your choosing. Inspect the resulting `groupCol` vector to make sure it has done what you expected it to.

2. We need to give the argument `col` a character vector of colour names, NOT a factor vector. Convert `groupCol` from a factor vector to a character vector. Inspect the resulting `groupCol` vector to make sure it has done what you expected it to.

*Hint*: see section\@ref(factors) to review the commands to convert between data types.

3. Use `plotMDS()` and the `col` argument to re-plot the MDS plot with phenotype colours.

### Plotting

There are two *phases* to making plots with base R: 

1. 'Initialising' a new plot, e.g. done with a 'plot' function like `plot()` or `plotMDS`.
2. Adding to an existing plot, e.g. adding a legend. 

We will explore adding base R plot annotation in the following challenge.

**Challenge 4.2**
1. Re-create the MDS plot with the sample names as labels and label coloured by phenotype group.

```{r}
groupCol <- group
levels(groupCol) <- c("blue", "green", "red")
groupCol <- as.character(groupCol)
groupCol
plotMDS(x, col = groupCol)
```

2. Add a title to your graph using the `main` argument to the `plotMDS()` function. Remember to give your title as a character.

3. Notice that the label of sample '9_6_5_11' is cut off. Amend this by changing the 'limits' of the x axis using the `xlim` argument to the `plotMDS()` function. `xlim` takes a numeric vector containing 2 elements, the min and max values of the x axis.

```{r}
plotMDS(x, col = groupCol, xlim = c(-2.8,4))
```

4. Add a legend to your plot using the `legend()` function. This function adds to an existing plot and thus should be run after the `plotMDS()` function. Useful arguments to `legend()` include:

* `x` - the position of the legend. It can be one of: "bottomright", "bottom", "bottomleft", "left", "topleft", "top", "topright", "right" or "center".
(technically you could also specify the position of the legend using `x` and `y` coordinates but we will not do this in the course)
* `legend` - a character vector specifying the text to display in the legend. Here it would be the names of the phenotype groups.
*Hint*: you can use function `levels()`, which provides categories of a factor vector.
* `text.col` - a character vector specifying the colours of the text in the legend. As there are 3 phenotype groups in our case, there should be 3 colours in this vector. These should also be the same colours as the labels within the plot.
*Hint*: the function `levels()` would also be useful for this argument.

## Linear modelling

**Challenge 4.3**



### Design matrix




## Statistical testing

**Challenge 4.4**
write top genes out to file

## Answers

**Challenge 4.1**

```{r}
#1 
levels(groupCol) <- c("blue", "green", "red")
groupCol
# 2
groupCol <- as.character(groupCol)
groupCol
# 3
plotMDS(x, labels = group, col = groupCol)
```

**Challenge 4.2**

```{r}
# 2
plotMDS(x, col = groupCol, xlim = c(-2.8,4))
```


**Challenge 4.3**

**Challenge 4.4**







