<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Chapter 3 Session 2 | Introduction to Base R</title>
  <meta name="description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor.">
  <meta name="generator" content="bookdown  and GitBook 2.6.7">

  <meta property="og:title" content="Chapter 3 Session 2 | Introduction to Base R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor." />
  <meta name="github-repo" content="lucyleeow/BaseR_Intro" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Session 2 | Introduction to Base R" />
  
  <meta name="twitter:description" content="Introductory tutorial to base R, with a biological focus. In particular it will focus on analysing RNA-sequencing data with Bioconductor." />
  

<meta name="author" content="Lucy Liu">


<meta name="date" content="2019-03-20">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="session-1.html">
<link rel="next" href="session-3.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to Base R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="session-1.html"><a href="session-1.html"><i class="fa fa-check"></i><b>2</b> Session 1</a><ul>
<li class="chapter" data-level="2.1" data-path="session-1.html"><a href="session-1.html#r-and-rstudio"><i class="fa fa-check"></i><b>2.1</b> R and RStudio</a></li>
<li class="chapter" data-level="2.2" data-path="session-1.html"><a href="session-1.html#help"><i class="fa fa-check"></i><b>2.2</b> Help</a></li>
<li class="chapter" data-level="2.3" data-path="session-1.html"><a href="session-1.html#r-scripts"><i class="fa fa-check"></i><b>2.3</b> R scripts</a></li>
<li class="chapter" data-level="2.4" data-path="session-1.html"><a href="session-1.html#working-directory"><i class="fa fa-check"></i><b>2.4</b> Working directory</a></li>
<li class="chapter" data-level="2.5" data-path="session-1.html"><a href="session-1.html#maths"><i class="fa fa-check"></i><b>2.5</b> Maths</a></li>
<li class="chapter" data-level="2.6" data-path="session-1.html"><a href="session-1.html#comps"><i class="fa fa-check"></i><b>2.6</b> Comparisons</a></li>
<li class="chapter" data-level="2.7" data-path="session-1.html"><a href="session-1.html#variables"><i class="fa fa-check"></i><b>2.7</b> Variables</a><ul>
<li class="chapter" data-level="2.7.1" data-path="session-1.html"><a href="session-1.html#errors-and-warnings"><i class="fa fa-check"></i><b>2.7.1</b> Errors and warnings</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="session-1.html"><a href="session-1.html#data-structures"><i class="fa fa-check"></i><b>2.8</b> Data structures</a><ul>
<li class="chapter" data-level="2.8.1" data-path="session-1.html"><a href="session-1.html#vector"><i class="fa fa-check"></i><b>2.8.1</b> Vector</a></li>
<li class="chapter" data-level="2.8.2" data-path="session-1.html"><a href="session-1.html#matrix"><i class="fa fa-check"></i><b>2.8.2</b> Matrix</a></li>
<li class="chapter" data-level="2.8.3" data-path="session-1.html"><a href="session-1.html#dataframe"><i class="fa fa-check"></i><b>2.8.3</b> Dataframe</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="session-1.html"><a href="session-1.html#reading-in-data"><i class="fa fa-check"></i><b>2.9</b> Reading in data</a></li>
<li class="chapter" data-level="2.10" data-path="session-1.html"><a href="session-1.html#data-types"><i class="fa fa-check"></i><b>2.10</b> Data types</a><ul>
<li class="chapter" data-level="2.10.1" data-path="session-1.html"><a href="session-1.html#factors"><i class="fa fa-check"></i><b>2.10.1</b> Factors</a></li>
</ul></li>
<li class="chapter" data-level="2.11" data-path="session-1.html"><a href="session-1.html#vectorisation"><i class="fa fa-check"></i><b>2.11</b> Vectorisation</a></li>
<li class="chapter" data-level="2.12" data-path="session-1.html"><a href="session-1.html#writing-out-data"><i class="fa fa-check"></i><b>2.12</b> Writing out data</a></li>
<li class="chapter" data-level="2.13" data-path="session-1.html"><a href="session-1.html#homework"><i class="fa fa-check"></i><b>2.13</b> Homework</a></li>
<li class="chapter" data-level="2.14" data-path="session-1.html"><a href="session-1.html#answers"><i class="fa fa-check"></i><b>2.14</b> Answers</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="session-2.html"><a href="session-2.html"><i class="fa fa-check"></i><b>3</b> Session 2</a><ul>
<li class="chapter" data-level="3.1" data-path="session-2.html"><a href="session-2.html#missing-values"><i class="fa fa-check"></i><b>3.1</b> Missing values</a></li>
<li class="chapter" data-level="3.2" data-path="session-2.html"><a href="session-2.html#subsetting"><i class="fa fa-check"></i><b>3.2</b> Subsetting</a><ul>
<li class="chapter" data-level="3.2.1" data-path="session-2.html"><a href="session-2.html#subsetting-with-indicies"><i class="fa fa-check"></i><b>3.2.1</b> Subsetting with indicies</a></li>
<li class="chapter" data-level="3.2.2" data-path="session-2.html"><a href="session-2.html#subsetting-with-logicals"><i class="fa fa-check"></i><b>3.2.2</b> Subsetting with logicals</a></li>
<li class="chapter" data-level="3.2.3" data-path="session-2.html"><a href="session-2.html#in"><i class="fa fa-check"></i><b>3.2.3</b> %in%</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="session-2.html"><a href="session-2.html#merge"><i class="fa fa-check"></i><b>3.3</b> Merge</a></li>
<li class="chapter" data-level="3.4" data-path="session-2.html"><a href="session-2.html#functions"><i class="fa fa-check"></i><b>3.4</b> Functions</a></li>
<li class="chapter" data-level="3.5" data-path="session-2.html"><a href="session-2.html#homework-1"><i class="fa fa-check"></i><b>3.5</b> Homework</a></li>
<li class="chapter" data-level="3.6" data-path="session-2.html"><a href="session-2.html#answers-1"><i class="fa fa-check"></i><b>3.6</b> Answers</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="session-3.html"><a href="session-3.html"><i class="fa fa-check"></i><b>4</b> Session 3</a><ul>
<li class="chapter" data-level="4.1" data-path="session-3.html"><a href="session-3.html#lists"><i class="fa fa-check"></i><b>4.1</b> Lists</a></li>
<li class="chapter" data-level="4.2" data-path="session-3.html"><a href="session-3.html#s4-objects"><i class="fa fa-check"></i><b>4.2</b> S4 objects</a></li>
<li class="chapter" data-level="4.3" data-path="session-3.html"><a href="session-3.html#dgelist"><i class="fa fa-check"></i><b>4.3</b> DGEList</a></li>
<li class="chapter" data-level="4.4" data-path="session-3.html"><a href="session-3.html#answers-2"><i class="fa fa-check"></i><b>4.4</b> Answers</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="session-4.html"><a href="session-4.html"><i class="fa fa-check"></i><b>5</b> Session 4</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Introduction to Base R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="session-2" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Session 2</h1>
<p>Last session we gained an understanding of R data structures and data types. In this session we will learn about:</p>
<ul>
<li>missing values</li>
<li>subsetting data structures</li>
<li>how to merge two dataframes</li>
<li>how functions are created and used</li>
</ul>
<div id="missing-values" class="section level2">
<h2><span class="header-section-number">3.1</span> Missing values</h2>
<p>First, let’s read in some data. Recall we use the function <code>read.delim()</code> and tell R not to label words as ‘Factors’ using <code>stringsAsFactors = FALSE</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/Ses2_genes.tsv&quot;</span>, <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>You can see in the environment tab that this dataframe has 200 rows (observations) and 5 columns (variables). You can also click on the <code>genes</code> entry in the Environment tab (red arrow in Figure <a href="session-2.html#fig:RS4">3.1</a>) to display the data in a new window (left):</p>
<div class="figure"><span id="fig:RS4"></span>
<img src="Images/RS4.png" alt="Viewing data from the 'Environment' tab." width="700px" />
<p class="caption">
Figure 3.1: Viewing data from the ‘Environment’ tab.
</p>
</div>
<p>Scrolling down the window, you can see that there are several <em>NA</em> entries. <code>NA</code> denotes a missing value in R.</p>
<p><code>NA</code>‘s have some interesting behaviour - they are ’contagious’. For example, if we wanted to take the mean of a vector numbers, which includes a <code>NA</code>, the answer is <code>NA</code>. If we think about it, it makes sense that the mean of two numbers and a ‘missing’ number that could be anything is <code>NA</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mean</span>(<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="ot">NA</span>))</code></pre></div>
<pre><code>## [1] NA</code></pre>
<p>Luckily many functions have a <code>na.rm</code> input, where you can tell it to remove <code>NA</code> values:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">mean</span>(<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">3</span>,<span class="ot">NA</span>), <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## [1] 2.5</code></pre>
<p>Later we will learn how to remove the <code>NA</code> values from our dataframe.</p>
</div>
<div id="subsetting" class="section level2">
<h2><span class="header-section-number">3.2</span> Subsetting</h2>
<p>Subsetting involves selecting a portion of a data structure and uses square brackets <code>[ ]</code>. There are two main ways to subset a data structure:</p>
<ul>
<li>Use indices - data structures are all ordered and ‘numbered’ in R. This means that you can refer to the 4th element in a vector or the element in the 1st row and 3rd column of a dataframe.</li>
<li>Use logicals - recall that a logical is <code>TRUE</code> or <code>FALSE</code>. You can use comparisons (e.g. equal to <code>==</code>, greater than <code>&gt;</code>) to determine if each element in your data structure meet your requirements and use this to subset your data.</li>
</ul>
<div id="subsetting-with-indicies" class="section level3">
<h3><span class="header-section-number">3.2.1</span> Subsetting with indicies</h3>
<p>Let’s take a look at subsetting vectors first. We can use <code>$</code> to obtain just one column from the <code>genes</code> dataframe. The output will be a vector.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes<span class="op">$</span>Count</code></pre></div>
<pre><code>##   [1]   979    15   932     0   214     2     2     0  1790    41  1033
##  [12]   700  2075   351    22     0     0   893     3   100 32100   269
##  [23]    34 20005  5751     0   275     9   849  7175   768   358   234
##  [34]  5065  2096  1994   757     0  2310     0     1     0  3091   810
##  [45]    30   816   817     1   580    33     0   941  1445   522     2
##  [56]  2346   572   431  5797     2    70  1236   966   235     2  3009
##  [67]   431   452  2602  3481    99 11857    35  1952     0   706  1317
##  [78]  1130    92     0   871   730  3892  3772    20     7     0   756
##  [89]     1   210   101    17    21   537  1240   930     1     1     0
## [100]   221  1599     0    67     1     0   315     0     0   739     0
## [111]     0   976     1   715 18636   289  1396     0   101   285  1665
## [122]    47     2  4483     2   363   169   135     0     2     0    11
## [133]  4088  1082    92     0  3888    17     0     0     0     0     0
## [144]     0     0     0     0     0  2442    30  1610     0     0     0
## [155]    16     1     2     0    47     0  1118     0     1     0     2
## [166]     3    22     0     0     0     0     0     0     0   431     0
## [177]     0    16     0     1     0     4     0     0     1     0     0
## [188]     0     0   301     0     0     0     0     0     0     0     0
## [199]     0    59</code></pre>
<p>We will assign this to a variable called <code>Counts</code>. <code>Counts</code> is now a variable that refers to a vector containing 200 integers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Counts &lt;-<span class="st"> </span>genes<span class="op">$</span>Count</code></pre></div>
<p>This will give you the 3rd element of the vector <code>Counts</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Counts[<span class="dv">3</span>]</code></pre></div>
<pre><code>## [1] 932</code></pre>
<p>This will give you all the elements from the 3rd to the 10th:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Counts[<span class="dv">3</span><span class="op">:</span><span class="dv">10</span>]</code></pre></div>
<pre><code>## [1]  932    0  214    2    2    0 1790   41</code></pre>
<p>This will give you the 3rd, 5th and 10th elements:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Counts[<span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">100</span>)]</code></pre></div>
<pre><code>## [1] 932 214 221</code></pre>
<p>Note that we have used <code>c()</code> within the square brackets. This is because R expects ONE ‘object’ within the square brackets. Thus, if you want to extract several indices, you must give it ONE vector containing the indices of the elements. A vector (e.g. <code>c(3,5,10)</code>) is considered one ‘object’ but the numbers <code>3,5,10</code> are considered three different ‘objects’.</p>
<p>Indeed, <code>3:10</code>, which we used earlier, is actually a vector of the numbers 3 to 10:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="dv">3</span><span class="op">:</span><span class="dv">10</span></code></pre></div>
<pre><code>## [1]  3  4  5  6  7  8  9 10</code></pre>
<p>Lastly, this gives you all the elements EXCEPT the elements 10 through to 200:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Counts[<span class="op">-</span>(<span class="dv">10</span><span class="op">:</span><span class="dv">200</span>)]</code></pre></div>
<pre><code>## [1]  979   15  932    0  214    2    2    0 1790</code></pre>
<p>Subsetting a dataframe is similar to subsetting a vector, except you now must specify which rows AND which columns you want. The syntax looks like this:</p>
<pre><code>genes[ (which rows you want) , (which columns you want) ]</code></pre>
<p>Within the square brackets, you must first tell R which rows you want LEFT of the comma, then which columns you want RIGHT of the comma.</p>
<p>The code below will give you the 3rd to 5th rows of the 2nd and 4th columns. Note that the output is a dataframe.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes[<span class="dv">3</span><span class="op">:</span><span class="dv">5</span>,<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">4</span>)]</code></pre></div>
<pre><code>##    SYMBOL GeneLength
## 3    Bmp2       2515
## 4 Cacna1e      12697
## 5    Ccnf       3099</code></pre>
<p>We could have also done this using the names of the columns. Note that column names are enclosed in quotes signifying that they are of the ‘character’ data type.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes[<span class="dv">3</span><span class="op">:</span><span class="dv">5</span>,<span class="kw">c</span>(<span class="st">&quot;SYMBOL&quot;</span>,<span class="st">&quot;GeneLength&quot;</span>)]</code></pre></div>
<pre><code>##    SYMBOL GeneLength
## 3    Bmp2       2515
## 4 Cacna1e      12697
## 5    Ccnf       3099</code></pre>
<p>If you leave the left side of comma empty, R will give you ALL the rows. If you leave the right side of the comma empty, R will give you ALL the columns.</p>
<p>This will give you the 2nd row and all the columns.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes[<span class="dv">2</span>,]</code></pre></div>
<pre><code>##   EntrezID SYMBOL TXCHROM GeneLength Count
## 2    11634   Aire   chr10       1929    15</code></pre>
</div>
<div id="subsetting-with-logicals" class="section level3">
<h3><span class="header-section-number">3.2.2</span> Subsetting with logicals</h3>
<p>If you recall from section <a href="session-1.html#comps">2.6</a>, you can make comparisons in R. The result of a comparison is either <code>TRUE</code> or <code>FALSE</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="dv">1</span> <span class="op">&lt;</span><span class="st"> </span><span class="dv">2</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>Comparisons are also vectorised:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes<span class="op">$</span>Count <span class="op">&lt;</span><span class="st"> </span><span class="dv">100</span></code></pre></div>
<pre><code>##   [1] FALSE  TRUE FALSE  TRUE FALSE  TRUE  TRUE  TRUE FALSE  TRUE FALSE
##  [12] FALSE FALSE FALSE  TRUE  TRUE  TRUE FALSE  TRUE FALSE FALSE FALSE
##  [23]  TRUE FALSE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE
##  [34] FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE  TRUE  TRUE FALSE FALSE
##  [45]  TRUE FALSE FALSE  TRUE FALSE  TRUE  TRUE FALSE FALSE FALSE  TRUE
##  [56] FALSE FALSE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE  TRUE FALSE
##  [67] FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE  TRUE FALSE FALSE
##  [78] FALSE  TRUE  TRUE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE FALSE
##  [89]  TRUE FALSE FALSE  TRUE  TRUE FALSE FALSE FALSE  TRUE  TRUE  TRUE
## [100] FALSE FALSE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE FALSE  TRUE
## [111]  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE
## [122]  TRUE  TRUE FALSE  TRUE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE
## [133] FALSE FALSE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [144]  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE FALSE  TRUE  TRUE  TRUE
## [155]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE
## [166]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE  TRUE
## [177]  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [188]  TRUE  TRUE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE
## [199]  TRUE  TRUE</code></pre>
<p>For each element in the vector <code>genes$Count</code>, R checks if it is less than 100, then returns either <code>TRUE</code> or <code>FALSE</code>. The output is a vector of logicals.</p>
<p>This can be used to subset in R. We will start with our <code>Counts</code> vector:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Counts[Counts <span class="op">&lt;</span><span class="st"> </span><span class="dv">100</span>]</code></pre></div>
<pre><code>##   [1] 15  0  2  2  0 41 22  0  0  3 34  0  9  0  0  1  0 30  1 33  0  2  2
##  [24] 70  2 99 35  0 92  0 20  7  0  1 17 21  1  1  0  0 67  1  0  0  0  0
##  [47]  0  1  0 47  2  2  0  2  0 11 92  0 17  0  0  0  0  0  0  0  0  0  0
##  [70] 30  0  0  0 16  1  2  0 47  0  0  1  0  2  3 22  0  0  0  0  0  0  0
##  [93]  0  0 16  0  1  0  4  0  0  1  0  0  0  0  0  0  0  0  0  0  0  0  0
## [116] 59</code></pre>
<p><code>Counts &lt; 100</code> will return a vector of 200 logicals, which indicate which elements are less than 100. Putting this inside square brackets will subset <code>Counts</code> such that only the elements that are less than 100 (the <code>TRUE</code>’s) remain.</p>
<p>This is done similarly in dataframes. The command below selects the rows where the <code>genes$Count</code> column equal to 0.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes[genes<span class="op">$</span>Count <span class="op">==</span><span class="st"> </span><span class="dv">0</span>,]</code></pre></div>
<p>Because there are so many rows with a <code>genes$Count</code> of 0, I’ll use the <code>head()</code> function to print out just the first 6 rows of the output:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(genes[genes<span class="op">$</span>Count <span class="op">==</span><span class="st"> </span><span class="dv">0</span>,])</code></pre></div>
<pre><code>##    EntrezID  SYMBOL TXCHROM GeneLength Count
## 4     12290 Cacna1e    chr1      12697     0
## 8     12647    Chat   chr14       2810     0
## 16    14800   Gria2    chr3       7749     0
## 17    14843    Gsx2    chr5       1665     0
## 26    17695    Msmb   chr14        522     0
## 38    19888     Rp1    chr1       9747     0</code></pre>
<p>If you add <code>!</code> to the start of the <code>genes$Count == 0</code> condition statement, you will get all the rows where <code>genes$Count</code> is NOT equal to 0.</p>
<p>Another way to think about it is that <code>genes$Count == 0</code> gives you a logical vector of 200 <code>TRUE</code>’s and <code>FALSE</code>’s and <code>!</code> flips everything such that the <code>TRUE</code>’s become <code>FALSE</code>’s and vice versa.</p>
<p>We do this here and print the first 6 rows of the output:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(genes[<span class="op">!</span><span class="st"> </span>genes<span class="op">$</span>Count <span class="op">==</span><span class="st"> </span><span class="dv">0</span>,])</code></pre></div>
<pre><code>##   EntrezID SYMBOL TXCHROM GeneLength Count
## 1    11480 Acvr2a    chr2       5681   979
## 2    11634   Aire   chr10       1929    15
## 3    12156   Bmp2    chr2       2515   932
## 5    12449   Ccnf   chr17       3099   214
## 6    12563   Cdh6   chr15       2632     2
## 7    12565   Cdh9   chr15       2903     2</code></pre>
<p><strong>Challenge 2.1</strong></p>
<ol style="list-style-type: decimal">
<li>Subset the <code>genes</code> dataframe to obtain the rows where the <code>Count</code> is less than or equal to 10 and the columns <code>TXCHROM</code> and <code>Count</code>.</li>
<li>The function <code>is.na()</code> checks if each element in a vector is <code>NA</code>:</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">is.na</span>(<span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">5</span>, <span class="ot">NA</span>))</code></pre></div>
<pre><code>## [1] FALSE FALSE  TRUE</code></pre>
<p>Use this function to subset the <code>genes</code> dataframe so that all rows where <code>TXCHROM</code> column is <code>NA</code> is removed.</p>
<ol start="3" style="list-style-type: decimal">
<li>Using the dataframe from above subset to get only the rows where the <code>TXCHROM</code> is ‘chr1’ and all columns.</li>
</ol>
<p><em>Hint</em>: you can refer to section <a href="session-1.html#comps">2.6</a> to check how to perform different types of comparisons in R.</p>
</div>
<div id="in" class="section level3">
<h3><span class="header-section-number">3.2.3</span> %in%</h3>
<p>In the last challenge, we used <code>==</code> to obtain the rows where <code>TXCHROM</code> is ‘chr1’. Another way to perform ‘matching’ tasks is with the <code>%in%</code> function.</p>
<p>The following command subsets the rows where <code>TXCHROM</code> is “chr1” or “chr2” and prints out the first 6 rows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(
  genes[genes<span class="op">$</span>TXCHROM <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;chr1&quot;</span>, <span class="st">&quot;chr2&quot;</span>),]
  )</code></pre></div>
<pre><code>##    EntrezID  SYMBOL TXCHROM GeneLength Count
## 1     11480  Acvr2a    chr2       5681   979
## 3     12156    Bmp2    chr2       2515   932
## 4     12290 Cacna1e    chr1      12697     0
## 18    15064     Mr1    chr1       2509   893
## 19    15112    Hao1    chr2       2029     3
## 21    15939    Ier5    chr1       3270 32100</code></pre>
<p><strong>Challenge 2.2</strong></p>
<p>There is an important difference between <code>==</code> and <code>%in%</code>.</p>
<p>Let’s start by creating a vector of numbers:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vect1 &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dv">10</span>,<span class="dv">10</span>,<span class="dv">5</span>,<span class="dv">5</span>,<span class="dv">8</span>,<span class="dv">8</span>)</code></pre></div>
<p>We check which elements in our vector is equal to 5.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vect1 <span class="op">==</span><span class="st"> </span><span class="dv">5</span></code></pre></div>
<pre><code>## [1] FALSE FALSE  TRUE  TRUE FALSE FALSE</code></pre>
<p>The output is what we would expect.</p>
<p>What if we wanted check which elements are equal to 5 OR 10? We might try something like this, where put the numbers we are checking for in a vector:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vect1</code></pre></div>
<pre><code>## [1] 10 10  5  5  8  8</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vect1 <span class="op">==</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">10</span>)</code></pre></div>
<pre><code>## [1] FALSE  TRUE  TRUE FALSE FALSE FALSE</code></pre>
<p>This isn’t the output we expected.</p>
<p>Let’s try the same task with <code>%in%</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vect1 <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">10</span>)</code></pre></div>
<pre><code>## [1]  TRUE  TRUE  TRUE  TRUE FALSE FALSE</code></pre>
<p>This output IS what we want.</p>
<p>Take a look at the code above and see if you can understand what <code>==</code> does and what <code>%in%</code> does.</p>
<p><em>Hint</em>: The story gets even more interesting if we try to use <code>==</code> to look for four numbers:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">vect1 <span class="op">==</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">5</span>,<span class="dv">10</span>,<span class="dv">1</span>,<span class="dv">3</span>)</code></pre></div>
<pre><code>## Warning in vect1 == c(5, 10, 1, 3): longer object length is not a multiple
## of shorter object length</code></pre>
<pre><code>## [1] FALSE  TRUE FALSE FALSE FALSE FALSE</code></pre>
<p>This warning message may seem a bit cryptic. The ‘longer’ object it is referring to is <code>vect1</code> which has 6 elements. The shorter object it is referring to is <code>c(5,10,1,3)</code>, which has 4 elements. Thus, it is saying that 6 is not a multiple of 4. The reason R wants the longer object to be a multiple of the shorter one, is key to understanding what is happening when we use <code>==</code>.</p>
</div>
</div>
<div id="merge" class="section level2">
<h2><span class="header-section-number">3.3</span> Merge</h2>
<p>Two dataframes can be combined with the <code>merge()</code> function.</p>
<p>First, let’s read in the file “Ses2_geneNames.tsv”, which contains gene names and their corresponding ENTREZIDs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene_names &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/Ses2_geneNames.tsv&quot;</span>, 
                         <span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>What we want to do now, is to add the <code>GENENAME</code> column from <code>gene_names</code> dataframe to the <code>genes</code> dataframe. We also want to make sure that when R merges the dataframes, the correct <code>GENENAME</code> is added to the correct row. You will notice that both the <code>genes</code> and <code>gene_names</code> dataframes have a column giving the EntrezIDs. Thus, we want <code>merge()</code> to match rows using EntrezIDS while it is merging.</p>
<p><code>merge()</code> has the following syntax:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">merge</span>(
  <span class="dt">x =</span>    <span class="co"># name of the first dataframe to merge</span>
  <span class="dt">y =</span>    <span class="co"># name of the second dataframe to merge</span>
  <span class="dt">by.x =</span> <span class="co"># name of the column to match, in the first dataframe</span>
  <span class="dt">by.y =</span> <span class="co"># name of the column to match in the second dataframe</span>
)</code></pre></div>
<p>Thus, to merge our two dataframes we can use:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes2 &lt;-<span class="st"> </span><span class="kw">merge</span>(<span class="dt">x =</span> genes, <span class="dt">y =</span> gene_names, 
                <span class="dt">by.x =</span> <span class="st">&quot;EntrezID&quot;</span>, <span class="dt">by.y =</span> <span class="st">&quot;ENTREZID&quot;</span>)

<span class="kw">head</span>(genes2)</code></pre></div>
<pre><code>##   EntrezID  SYMBOL TXCHROM GeneLength Count
## 1    11480  Acvr2a    chr2       5681   979
## 2    11634    Aire   chr10       1929    15
## 3    12156    Bmp2    chr2       2515   932
## 4    12290 Cacna1e    chr1      12697     0
## 5    12449    Ccnf   chr17       3099   214
## 6    12563    Cdh6   chr15       2632     2
##                                                                                GENENAME
## 1                                                                  activin receptor IIA
## 2 autoimmune regulator (autoimmune polyendocrinopathy candidiasis ectodermal dystrophy)
## 3                                                          bone morphogenetic protein 2
## 4                          calcium channel, voltage-dependent, R type, alpha 1E subunit
## 5                                                                              cyclin F
## 6                                                                            cadherin 6</code></pre>
<p><strong>Challenge 2.3</strong></p>
<p>You may have noticed that there are 200 rows in the <code>genes</code> dataframe but 290 rows in the <code>gene_names</code> dataframe. This means that there are more gene names contained. We can also specify which rows to keep when merging:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">merge</span>(
  <span class="dt">x =</span>     <span class="co"># name of the first dataframe to merge</span>
  <span class="dt">y =</span>     <span class="co"># name of the second dataframe to merge</span>
  <span class="dt">by.x =</span>  <span class="co"># name of the column to match, in the first dataframe</span>
  <span class="dt">by.y =</span>  <span class="co"># name of the column to match in the second dataframe</span>
  <span class="dt">all.x =</span> <span class="co"># logical. If TRUE, keep all rows from the first dataframe, </span>
          <span class="co"># even if does not have a matching row in the second dataframe</span>
  <span class="dt">all.y =</span> <span class="co"># logical. If TRUE, keep all rows from the second dataframe, </span>
          <span class="co"># even if does not have a matching row in the second dataframe</span>
)</code></pre></div>
<p>Merge the two dataframes again, but this time keep all rows from both dataframes.</p>
</div>
<div id="functions" class="section level2">
<h2><span class="header-section-number">3.4</span> Functions</h2>
<p>A function, as the namne suggests performs a function. We have already used many functions. For example, the <code>read.delim()</code> function reads in data, the <code>sum()</code> function adds numbers up and the <code>merge()</code> function above merges two dataframes.</p>
<p>Formally, when using a function, brackets (<code>( )</code>) are always included after the name of the function. Within the brackets is where input (technically ‘arguments’) to the function goes. You can find out what inputs an argument takes by looking at the help file.</p>
<p>For example, let’s take a look at the <code>rnorm()</code> function help file. This function generates a random number from a normal distribution.</p>
<div class="figure" style="text-align: center"><span id="fig:rnorm"></span>
<img src="Images/rnorm_help.png" alt="rnorm help file." width="600px" />
<p class="caption">
Figure 3.2: rnorm help file.
</p>
</div>
<p>We can see from the help file that <code>rnorm()</code> takes in 3 arguments:</p>
<pre><code>rnorm(n, mean = 0, sd = 1)</code></pre>
<p>The help file also tells us what each of these arguments does:</p>
<pre><code>n   - number of observations. If length(n) &gt; 1, the length is taken to be the number required.
mean - vector of means.
sd - vector of standard deviations.</code></pre>
<p><strong>Challenge 2.4</strong></p>
<ol style="list-style-type: decimal">
<li>Take a look at what the <code>rnorm()</code> function outputs in the code below and have a go plugging various inputs to the function. Then try to answer the following questions:</li>
</ol>
<ul>
<li>What does <code>rnorm()</code> do when you only give it one number to the <code>mean</code> and <code>sd</code> arguments?</li>
<li>What happens when you give either argument a vector of numbers?</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rnorm</span>(<span class="dv">4</span>, <span class="dt">mean =</span> <span class="dv">10</span>, <span class="dt">sd =</span> <span class="dv">1</span>)</code></pre></div>
<pre><code>## [1]  9.524970 10.401953 10.202806  9.548079</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rnorm</span>(<span class="dv">3</span>, <span class="dt">mean =</span> <span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">0</span>), <span class="dt">sd =</span> <span class="dv">1</span>)</code></pre></div>
<pre><code>## [1] 100.6871521  -0.9455308 100.4756134</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>Take a look at the following code:</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rnorm</span>(<span class="dv">3</span>,<span class="dv">1</span>,<span class="dv">10</span>)</code></pre></div>
<pre><code>## [1]  -1.348913   7.631334 -11.347598</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rnorm</span>(<span class="dv">3</span>,<span class="dv">10</span>,<span class="dv">1</span>)</code></pre></div>
<pre><code>## [1] 10.11129  9.26211  9.58699</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rnorm</span>(<span class="dv">3</span>, <span class="dt">mean =</span> <span class="dv">10</span>, <span class="dt">sd =</span> <span class="dv">0</span>)</code></pre></div>
<pre><code>## [1] 10 10 10</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rnorm</span>(<span class="dv">3</span>, <span class="dt">sd =</span> <span class="dv">10</span>, <span class="dt">mean =</span> <span class="dv">0</span>)</code></pre></div>
<pre><code>## [1] -1.174505 27.074352  2.006016</code></pre>
<p>How does the order of the arguments you input to the function affect the output? What happens when you include the name of the argument before your input?</p>
<p>Note that as the function generates a random number, the numbers you will get from running the function will be different to the ones generated above.</p>
</div>
<div id="homework-1" class="section level2">
<h2><span class="header-section-number">3.5</span> Homework</h2>
<p>We will use the Bioconductor packages ‘limma’ and ‘edgeR’ in the next session. Before the next sesssion, please install these packages with the following commands:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">&quot;BiocManager&quot;</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
    <span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)
BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="kw">c</span>(<span class="st">&quot;edgeR&quot;</span>, <span class="st">&quot;limma&quot;</span>), <span class="dt">version =</span> <span class="st">&quot;3.8&quot;</span>)</code></pre></div>
</div>
<div id="answers-1" class="section level2">
<h2><span class="header-section-number">3.6</span> Answers</h2>
<p><strong>Challenge 2.1</strong></p>
<ol style="list-style-type: decimal">
<li>This code obtains rows where <code>Count</code> is less than or equal to 10 and prints the first 6 rows using <code>head()</code>:</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(genes[genes<span class="op">$</span>Count <span class="op">&lt;=</span><span class="st"> </span><span class="dv">10</span>,])</code></pre></div>
<pre><code>##    EntrezID  SYMBOL TXCHROM GeneLength Count
## 4     12290 Cacna1e    chr1      12697     0
## 6     12563    Cdh6   chr15       2632     2
## 7     12565    Cdh9   chr15       2903     2
## 8     12647    Chat   chr14       2810     0
## 16    14800   Gria2    chr3       7749     0
## 17    14843    Gsx2    chr5       1665     0</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>This code removes all rows where the <code>TXCHROM</code> column has a NA value (keeping all columns) and prints the first 6 rows:</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(genes[<span class="op">!</span><span class="st"> </span><span class="kw">is.na</span>(genes<span class="op">$</span>TXCHROM),])</code></pre></div>
<pre><code>##   EntrezID  SYMBOL TXCHROM GeneLength Count
## 1    11480  Acvr2a    chr2       5681   979
## 2    11634    Aire   chr10       1929    15
## 3    12156    Bmp2    chr2       2515   932
## 4    12290 Cacna1e    chr1      12697     0
## 5    12449    Ccnf   chr17       3099   214
## 6    12563    Cdh6   chr15       2632     2</code></pre>
<ol start="3" style="list-style-type: decimal">
<li>First we save the dataframe from the above as a variable called <code>genes_noNA</code>, then we subset to get only the rows where <code>TXCHROM</code> is ‘chr1’. Again we only print the first 6 rows of this output.</li>
</ol>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes_noNA &lt;-<span class="st"> </span>genes[<span class="op">!</span><span class="st"> </span><span class="kw">is.na</span>(genes<span class="op">$</span>TXCHROM),]

<span class="kw">head</span>(genes_noNA[genes_noNA<span class="op">$</span>TXCHROM <span class="op">==</span><span class="st"> &quot;chr1&quot;</span>,])</code></pre></div>
<pre><code>##    EntrezID  SYMBOL TXCHROM GeneLength Count
## 4     12290 Cacna1e    chr1      12697     0
## 18    15064     Mr1    chr1       2509   893
## 21    15939    Ier5    chr1       3270 32100
## 31    18777  Lypla1    chr1       2433   768
## 36    19775    Xpr1    chr1       7651  1994
## 38    19888     Rp1    chr1       9747     0</code></pre>
<p><strong>Challenge 2.2</strong></p>
<p>What <code>==</code> does is compare <code>vect1</code> with <code>c(5,10)</code> one by one. Since <code>c(5,10)</code> only has two elements, R repeats this shorter vector until it is the same length as the longer vector. This is called ‘recycling’.</p>
<p>Thus, the comparisons being made is the top row with each corresponding value in the bottom row, with the result being shown in the third row:</p>
<pre><code>10     10    5    5     8     8

5      10    5    10    5    10 

FALSE TRUE TRUE FALSE FALSE FALSE</code></pre>
<p>R gives you a warning whenever the length of the shorter vector is NOT a multiple of the longer vector.</p>
<p>When there were 4 elements in the shorter vector, it was not repeated a whole number of times - it was repeated 1.5 times. The comparisons being made between <code>vect1</code> and <code>c(5,10,1,3)</code> are:</p>
<pre><code>10     10    5    5     8     8

5      10    1    3     5    10 

FALSE TRUE FALSE FALSE FALSE FALSE</code></pre>
<p><code>%in%</code> simply performs matching and does not take order of the two vectors into consideration. It checks whether the values in <code>vect1</code> matches either number in <code>c(5,10)</code>. It thus gives us the result we expect.</p>
<p><strong>Challenge 2.3</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes2 &lt;-<span class="st"> </span><span class="kw">merge</span>(<span class="dt">x =</span> genes, <span class="dt">y =</span> gene_names, 
                <span class="dt">by.x =</span> <span class="st">&quot;EntrezID&quot;</span>, <span class="dt">by.y =</span> <span class="st">&quot;ENTREZID&quot;</span>,
                <span class="dt">by.x =</span> <span class="ot">TRUE</span>, <span class="dt">by.y =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><strong>Challenge 2.4</strong></p>
<ol style="list-style-type: decimal">
<li>If you provide <code>rnorm()</code> with a vector of inputs to either the <code>mean</code> or <code>sd</code> argument, it will use each element in that vector for successive random numbers generated and recycle the vector if it is shorter than the number of random numbers required.</li>
</ol>
<p>Thus, the following code:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rnorm</span>(<span class="dv">3</span>, <span class="dt">mean =</span> <span class="kw">c</span>(<span class="dv">100</span>,<span class="dv">0</span>), <span class="dt">sd =</span> <span class="dv">1</span>)</code></pre></div>
<p>will generate 3 random numbers from a normal distribution of:</p>
<pre><code>1. mean of 100 and sd of 1
2. mean of 0 and sd of 1
3. mean of 100 and sd of 1</code></pre>
<ol start="2" style="list-style-type: decimal">
<li>If you do not provide the name of the argument, <code>rnorm()</code> will use the first number provided as the argument to <code>n</code>, the second number provided as the input to <code>mean</code> and the third number provided as the input to <code>sd</code>.</li>
</ol>
<p>Thus, <code>rnorm(3,1,10)</code> generates 3 random numbers from a normal distribution with a mean of 1 and a standard deviation of 10 and <code>rnorm(3,10,1)</code> will generate 3 random numbers from a normal distribution with a mean of 10 and a standard deviation of 1.</p>
<p>If you give the argument name with the input, it does not matter what order you provide the inputs.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="session-1.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="session-3.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/rstudio/bookdown-demo/edit/master/02-session2.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"download": ["IntroBaseR.pdf", "IntroBaseR.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
